File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque
def solution(n, wires):
    answer = n-2
    graph = [[] for _ in range(n+1)]
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    def bfs(start):
        visited = [0] * (n+1)
        q = deque([start])
        visited[start]=1
        cnt = 1
        while q:
            s = q.popleft()
            for i in graph[s]:
                if not visited[i]:
                    q.append(i)
                    visited[i] = 1
                    cnt+=1
        return cnt
    answer = n
    for a, b in wires:
        graph[a].remove(b)
        graph[b].remove(a)
        
        answer = min(abs(bfs(a)-bfs(b)), answer)
        
        graph[a].append(b)
        graph[b].append(a)
        
    return answer

File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public int solution(int n, int[][] wires) {
        		int answer = n-2;
		ArrayList<Integer>[] graph = new ArrayList[n+1];
		
		for (int i=0; i<n+1; i++) {
			graph[i] = new ArrayList<>();
		}
		
		for (int i=0; i<wires.length; i++) {
			graph[wires[i][0]].add(wires[i][1]);
			graph[wires[i][1]].add(wires[i][0]);
		}
		
//		for (int i=0; i<n+1; i++) {
//			System.out.print(i+": ");
//			for (int x: graph[i])
//				System.out.print(x+" ");
//			System.out.println();
//		}
		
		for (int i=0; i<wires.length; i++) {
			int a = wires[i][0];
			int b = wires[i][1];
			graph[a].remove(graph[a].indexOf(b));
			graph[b].remove(graph[b].indexOf(a));
			
			answer = Math.min(Math.abs(bfs(a, graph, n)-bfs(b, graph, n)), answer);
			
			graph[a].add(b);
			graph[b].add(a);
		}
        return answer;
    }
    static int bfs(int start, ArrayList<Integer>[] graph, int n) {
		boolean[] visited = new boolean[n+1];
		Queue<Integer> q = new LinkedList<>();
		q.offer(start);
		visited[start] = true;
		int cnt=1;
		while (!q.isEmpty()) {
			int node = q.poll();
			for (int point: graph[node]) {
				if (!visited[point]) {
					q.offer(point);
					visited[point] = true;
					cnt++;
				}
			}
		}
//		System.out.println("start: "+start+" cnt: "+cnt);
		return cnt;
		
	}
}